agent/common.py

```diff
@@ -0,0 +1,276 @@
+    return ip, port
+
+
+def _is_ipv6(potential_ip: str) -> bool:
+    """check if the potential_ip is a valid ipv6.
+
+    Args:
+        potential_ip: string.
+
+    Returns:
+        - boolean.
+    """
+    try:
+        ipaddress.IPv6Address(potential_ip)
+        return True
+    except ValueError:
+        return False
+
+
+def build_vuln_location(
+    target_url: str | None,
+    file_path: str | None = None,
+    file_content: bytes | None = None,
+    selector: str | None = None,
+    content_url: str | None = None,
+) -> agent_report_vulnerability_mixin.VulnerabilityLocation | None:
+    """Build VulnerabilityLocation based on the asset.
+
+    Args:
+        target_url: The URL
+        file_path: The path of the file
+        file_content: The content of the file
+        selector: The message selector
+        content_url: The content url of the file
+    Returns:
+        The vulnerability location object.
+    """
+
+    asset: (
+        ipv4_asset.IPv4
+        | ipv6_asset.IPv6
+        | domain_asset.DomainName
+        | file_asset.File
+        | android_aab_asset.AndroidAab
+        | android_apk_asset.AndroidApk
+        | ios_ipa_asset.IOSIpa
+    )
+
+    if selector == "v3.asset.file.android.apk":
+        asset = android_apk_asset.AndroidApk(
+            path=file_path, content=file_content, content_url=content_url
+        )
+        return agent_report_vulnerability_mixin.VulnerabilityLocation(
+            asset=asset,
+            metadata=[
+                agent_report_vulnerability_mixin.VulnerabilityLocationMetadata(
+                    metadata_type=agent_report_vulnerability_mixin.MetadataType.FILE_PATH,
+                    value=file_path or "",
+                )
+            ],
+        )
+    elif selector == "v3.asset.file.android.aab":
+        asset = android_aab_asset.AndroidAab(
+            path=file_path, content=file_content, content_url=content_url
+        )
+        return agent_report_vulnerability_mixin.VulnerabilityLocation(
+            asset=asset,
+            metadata=[
+                agent_report_vulnerability_mixin.VulnerabilityLocationMetadata(
+                    metadata_type=agent_report_vulnerability_mixin.MetadataType.FILE_PATH,
+                    value=file_path or "",
+                )
+            ],
+        )
+    elif selector == "v3.asset.file.ios.ipa":
+        asset = ios_ipa_asset.IOSIpa(
+            path=file_path, content=file_content, content_url=content_url
+        )
+        return agent_report_vulnerability_mixin.VulnerabilityLocation(
+            asset=asset,
+            metadata=[
+                agent_report_vulnerability_mixin.VulnerabilityLocationMetadata(
+                    metadata_type=agent_report_vulnerability_mixin.MetadataType.FILE_PATH,
+                    value=file_path or "",
+                )
+            ],
+        )
+    elif selector == "v3.asset.file":
+        asset = file_asset.File(path=file_path, content=file_content)
```

Comment:
For the file case, you check the `metadata_oneof` field & create the corresponding asset from it.